// domain/lib/Domain.js
module.exports = Domain;

var q = require("q");

var Repository = require("./Repository"),
	is = require("./_type");

if (typeof window !== 'undefined') {
	Emitter = require("emitter")
} else {
	Emitter = require("events").EventEmitter;
}

function Domain() {

	if (!(this instanceof Domain)) {
		return new Domain();
	}

	var self = this,
		emitter = new Emitter,
		AggreClassList = [],
		repositoryList = [],
		commandHandleList = [],
		serviceList = [],
		get = null,
		isSeal = false,
		Aggres = {},
		repos = {},
		openMethods = [],
		services = {},
		commandHandles = {},
		publish = function() {
			emitter.emit.apply(emitter, arguments);
		}

		// Only for testing.
	this._my = {
		Repository: Repository,
		Aggres: Aggres,
		services: services,
		repos: repos
	};

	function _push(cache, o) {
		if (is(o) === "array") {
			o.forEach(function(obj) {
				cache.push(obj);
			})
		} else {
			cache.push(o);
		}
	}

	function _register(type, o) {
		switch (type) {
			case "get":
				get = o;
				break;
			case "AggreClass":
				_push(AggreClassList, o);
				break;
			case "repository":
				_push(repositoryList, o)
				break;
			case "commandHandle":
				_push(commandHandleList, o);
				break;
			case "service":
				_push(serviceList, o);
				break;
			case "listener":
				if (is(o) === "array") {
					o.forEach(function(obj) {
						var handles = obj({
							q:q,
							repos: repos,
							services: services
						});
						if (is(handles) === "array") {
							handles.forEach(function(handle) {
								emitter.on(handle.eventName, handle);
							})
						} else {
							emitter.on(handles.eventName, handles);
						}
					})
				} else {
					var handles = o({
						q:q,
						repos: repos,
						services: services
					});
					if (is(handles) === "array") {
						handles.forEach(function(handle) {
							emitter.on(handle.eventName, handle);
						})
					} else {
						emitter.on(handles.eventName, handles);
					}
				}
				break;
			case "listenerOne":
				if (is(o) === "array") {
					o.forEach(function(obj) {
						var handles = obj({
							q:q,
							repos: repos,
							services: services
						});
						if (is(o) === "array") {
							handle.forEach(function(handle) {
								emitter.once(handle.eventName, handle);
							})
						} else {
							emitter.once(handles.eventName, handles);
						}
					})
				} else {
					var handles = o({
						q:q,
						repos: repos,
						services: services
					});
					if (is(handles) === "array") {
						handles.forEach(function(handle) {
							emitter.once(handle.eventName, handle);
						})
					} else {
						emitter.once(handles.eventName, handles);
					}
				}
				break;
		}
	}

	this.register = function() {

		if (isSeal) return this;

		var self = this;

		var go = true,
			type = null;

		while (go) {
			var first = [].shift.call(arguments);
			if (is(first) === "string") {
				type = first;
				var second = [].shift.call(arguments);
				_register(first, second);
			} else if (!first) {
				go = false;
			} else {
				_register(type, first);
			}
		}

		return this;
	}

	this.seal = function() {

		if (isSeal) {
			return this;
		} else {
			isSeal = true;
		}
		AggreClassList.forEach(function(wrap) {
			var o = wrap({
				q:q,
				repos: repos,
				services: services,
				publish: publish
			});
			if (is(o) === "array") {
				o.forEach(function(a) {
					Aggres[a.className] = a;
				})
			} else {
				Aggres[o.className] = o;
			}
		})

		serviceList.forEach(function(wrap) {
			var o = wrap({
				q:q,
				repos: repos,
				services: services
			});
			if (is(o) === "array") {
				o.forEach(function(a) {
					services[a.serviceName] = a;
				})
			} else {
				services[o.serviceName] = o;
			}
		})

		Repository.prototype._get = get;
		Repository.prototype._publish = publish;


		repositoryList.forEach(function(wrap) {
			var o = wrap({
				q:q,
				Repository: Repository,
				Aggres: Aggres,
				services: services
			});
			if (is(o) === "array") {
				o.forEach(function(a) {
					repos[a.className] = a;
				})
			} else {
				repos[o.className] = o;
			}
		})

		commandHandleList.forEach(function(wrap) {
			var o = wrap({
				q:q,
				repos: repos,
				services: services
			});
			if (is(o) === "array") {
				o.forEach(function(a) {
					commandHandles[a.commandName] = a;
				})
			} else {
				commandHandles[o.commandName] = o;
			}
		})
		return this;
	}

	/*
	 *@param commandName , command name.
	 *@param args , command handle's arguments.
	 *@param callback , command handle's callback.
	 */
	this.exec = function(commandName, args, callback) {
		args = args || {};
		callback = callback || function fn(){};
		var handle = commandHandles[commandName];
		handle(args, callback);
	}

	this.openMethod = function() {
		if (isSeal) return this;
		for (var i = 0; i < arguments.length; i++) {
			openMethods.push(arguments[i]);
		}
		return this;
	}

	/*
	 * @api public
	 * only call aggre object method, no result.
	 * @param methodName , AggreName.methodName
	 * @param id , aggre's ID.
	 * @param args , method arguments array.
	 * @param [callback] callback(err) mean, if there is an error.
	 * @return , if params error , then return ParamError object (see strict-method).
	 * @throw if params error.
	 */
	this.call = function(methodName, id, args, callback) {
		
		args = args || [];
		callback = callback || function fn(){};
		
		if (openMethods.indexOf(methodName) !== -1) {
			try {
				var cm = methodName.split("."),
					className = cm[0],
					methodName = cm[1],
					repo = repos[className];
				repo.get(id, function(err, a) {
					try {
						var method = a[methodName];
						var rs = method.apply(a, args);
						if(q.isPromise(rs)){
							rs.then(callback);
						}else{
							callback(rs);
						}
					} catch (err) {
						callback(err);
					}
				});
			} catch (err) {
				callback(err);
			}
		}else{
			callback("no method");
		}
		
	}

	this.on = function(event, listener) {
		emitter.on(event, listener);
	}

	this.once = function(event, listener) {
		emitter.once(event, listener);
	}

	this.removeListener = function(event, listener) {
		if (typeof window !== 'undefined') {
			emitter.off(event, listener);
		} else {
			emitter.removeListener(event, listener);
		}
	}

}

module.exports = Domain;
