// domain/lib/Repository.js
if (typeof window !== "undefined") {
    uuid = require("uuid");
} else {
    uuid = require("node-uuid").v1;
}

function Repository(className) {

    if(!className) throw new Error("No class name");
    
    var cache = {},self = this;
    this.className = className;

    this.getData = function(id, callback) {
        var self = this;
        this.get(id, function(err, aggre) {
            if (aggre) {
                callback(self._aggre2data(aggre));
            } else {
                callback();
            }
        })
    }

    /* @public */
    this.get = function(id, callback) {
        var self = this;
        if (cache[id]) {
            callback(undefined, cache[id])
        } else if (typeof cache[id] === 'boolean') {
            callback();
        } else {
            self._get(this.className, id, function(err, data) {
                if (data) {
                    var aggobj = cache[id] = self._data2aggre(data);
                    callback(undefined, aggobj);
                } else {
                    callback(err);
                }
            })
        }
    }

    /* @public */
    this.create = function(data, callback) {
        var self = this;
        this._create(data, function(err, aggobj) {
            if (err) {
                callback(err);
            } else {
                var data = self._aggre2data(aggobj);
                if (!data.id) {
                    data.id = uuid();
                }
                aggobj = self._data2aggre(data);
                cache[data.id] = aggobj;
                callback(undefined, aggobj);
                self._publish(className + ".*.create", className, data);
                self._publish("*.*.create", className, data);
            };
        });
    }

    /* @public */
    this.remove = function(id) {
        cache[id] = false;
        this._publish(this.className + "." + id + ".remove", className, id);
        this._publish(this.className + ".*.remove", className, id);
        this._publish("*.*.remove", className, id);
    }

    // DOTO test step.
    this.clearTime = 1000 * 60 * 5;
    (function loopClear() {
        setTimeout(function() {
            cache = {}
            loopClear();
        }, self.clearTime);
    }());

}

module.exports = Repository;
